### ISSUE TITLE
Unexpected error (code 18) on session start

### ISSUE THREAD

**Issue by lordlycastle** [Azure/azure-spatial-anchors-samples#103]
I get the following meaningful message: `Error code: 18. Error message: Unexpected error gathering feedback data` when I start a session. It is logged in Error event. All I do is start the session and then that starts pouring.

Here's the script I use to control the anchor manager: https://gist.github.com/lordlycastle/4c440e9df9ad480c1ed2b6688269a6b6

All I do is call `QuickStartSession`. Happens with both 1.3.3 and 2.0.0.

**Comment by darax** [comment]
Hi @lordlycastle,

Do the default demos work?  

I understand that there is a huge spew of messages, could there be a different message hidden in that list that gives a clue?  What's happening behind the scenes is ASA is trying to figure out how much computer vision coverage the device has, and getting that information is failing in a way that we aren't currently handling/haven't seen before.

Does QuickStartSession->CheckManagerExists end up creating the SpatialAnchorManager? If it does, QuickStartSession() would add the SpatialAnchorManager, Unity would call SpatialAnchorManager.Awake() function, and then QuickStartSession() would call the SpatialAnchorManager.StartSession() function before Unity called the SpatialAnchorManager.Start() function that sets everything up.  An idea to avoid this would be to make CheckManagerExists return an async Task, and after adding the SpatialAnchorManager component, you could yield Task.Delay(100) to give Unity a chance to call SpatialAnchorManager.Stat().

**Comment by lordlycastle** [comment]
That’s not it. I always had an instance of manager in scene and quick start wasn’t called on awake from some other script but when I pressed a button. So it had enough time to initialize.

**Comment by lordlycastle** [comment]
The logs had nothing useful that caught my eye. It was just version numbers of sdk etc.

**Comment by darax** [comment]
What platform are you testing on?  I'll see if I can reproduce this with your script.

**Comment by lordlycastle** [comment]
I am on HL2.

**Comment by darax** [comment]
I think I see the problem now. 

in AzureSpatialANchorsManager.CloudManager_SessionUpdated 
args.Status is being treated as an enum, when it is a class.  I patched this issue on my end, and things seemed to start working.

**Comment by lordlycastle** [comment]
How would that cause this problem? Why would it always be equal to 18 raw value?

I think that’s a reminiscence from 1.3.3 version as I tried to get this working on that after it failed on 2.0. 

And since SessionStatus is a class type and not EnumType GetName function would throw an exception according to the docs. https://docs.microsoft.com/en-us/dotnet/api/system.enum.getname 

I’ll give it a try. What version are you using of SDK?

**Comment by lordlycastle** [comment]
Also the message I’ve included in the OP is from the error event not from session updated event where that class is used as Enum. 

See the log format from message in CloudManager. 

 So I’m even less convinced now that was the issue.

**Comment by darax** [comment]
I'm assuming that you tried to patch the callback function and are still getting an error.

I'll explain my reasoning.

When SpatialAnchorManager starts the session, the first thing it does is send a signal to the ASA SDK to update the session status. As a result, ASA calls the callback associated with session updates.  In this case AzureSpatialAnchorsManager.CloudManager_SessionUpdated gets the callback and misinterprets the args.Status value as an enum. As documented, this raises an exception.  This exception is being raised on the ASA thread that is calling the callback. The ASA SDK catches the exception, but doesn't handle or try to understand exceptions raised from the callback.  It does send the 'unexpected error' to the 'Error' delegate.  This process repeats periodically, with every subsequent attempt to update the application with the session status results in the same sequence of events.  StatusCallback->Exception->CaughtException->ErrorCallback.

I'm using 2.1.1. The error code is different ("Unknown" vs 18).  I tested on Android using Unity's script debugging to step through the 'QuickStartSession' and hit the issue. I just tried this on my HL2 and got the same results.  I integrated AzureSpatialAnchorsManager.cs into the basic demo scene and (after patching) was able to get as far as locating an anchor I just placed. There is another Enum.GetName issue in the CloudManager_AnchorLocated callback (SessionStatus used when LocateAnchorStatus was the correct type).
